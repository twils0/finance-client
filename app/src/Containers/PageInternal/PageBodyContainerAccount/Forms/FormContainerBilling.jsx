import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { StripeProvider, Elements } from 'react-stripe-elements';
import { BeatLoader } from 'halogenium';

import theme from '../../../../themes';

import { requestStatusTypes } from '../../../../Constants/universalConstants';
import { statusNames, fieldNames } from '../../../../Constants/dataConstantsAccount';
import { errorMessages } from '../../../../Constants/uiConstantsApp';
import { formNames, inputNames } from '../../../../Constants/uiConstantsAccount';
import { setInputValueError } from '../../../../Actions/uiActionsAccount';

import FlexColumn from '../../../../Components/Flex/FlexColumn';
import FormBilling from '../../../../Components/PageInternal/PageBodyAccount/Forms/FormBilling';

class FormContainerBilling extends React.Component {
  componentWillMount() {
    const { inputs, dataAccountFields, handleInputValueError } = this.props;

    const uiNameOnCard = inputs[inputNames[formNames.BILLING].NAME_ON_CARD];
    const uiPromoCode = inputs[inputNames[formNames.BILLING].PROMO_CODE];
    // uiStripe.value is used for error handling
    // do not set it equal to dataStripe

    uiNameOnCard.value = dataAccountFields[fieldNames.NAME_ON_CARD].value;
    uiPromoCode.value = dataAccountFields[fieldNames.PROMO_CODE].value;
    if (
      dataAccountFields[fieldNames.PROMO_CODE].value &&
      !dataAccountFields[fieldNames.PROMO_CODE_VALID].value
    ) {
      uiPromoCode.errorMessage = errorMessages.PROMO_CODE_EXPIRED;
    }

    handleInputValueError(uiNameOnCard);
    handleInputValueError(uiPromoCode);
  }

  componentDidMount() {
    this.props.setHeight(formNames.BILLING);
  }

  componentWillUpdate(nextProps) {
    if (
      nextProps.statusAccount[statusNames.UPDATE_STRIPE_FIELDS_REQUEST].status ===
      requestStatusTypes.LOADING
    ) {
      const { clearElement, resetStripeElement } = nextProps;

      clearElement(formNames.BILLING);
      resetStripeElement();
    }
  }

  componentDidUpdate() {
    this.props.setHeight(formNames.BILLING);
  }

  componentWillUnmount() {
    this.props.clearElement(formNames.BILLING);
  }

  handleInputChange = (inputEvent) => {
    inputEvent.preventDefault();
    const { inputs, handleInputValueError } = this.props;
    const input = { ...inputs[inputEvent.target.id] };

    if (input) {
      if (inputEvent.target.value || input.value) {
        input.value = inputEvent.target.value;
      }

      if (input.errorMessage) {
        switch (input.errorMessage) {
          case errorMessages.NO_NAME_ON_CARD: {
            if (input.value) {
              input.errorMessage = '';
            }
            break;
          }
          case errorMessages.PROMO_CODE_INVALID: {
            input.errorMessage = '';
            break;
          }
          case errorMessages.PROMO_CODE_EXPIRED: {
            input.errorMessage = '';
            break;
          }
          default:
            input.errorMessage = '';
            break;
        }
      }
    }

    handleInputValueError(input);
  };

  render() {
    const {
      stripeRef,
      heightRef,
      stripe,
      rowWidthIndent,
      headerWidthBufferRight,
      mediaWidth,
      height,
      rowHeight,
      rowHeightBuffer,
      statusAccount,
      currentForm,
      editForm,
      inputs,
    } = this.props;

    if (
      statusAccount[statusNames.UPDATE_STRIPE_FIELDS_REQUEST].status === requestStatusTypes.LOADING
    ) {
      const heightString = `${height}px`;

      return (
        <FlexColumn width="100%" height={heightString}>
          <BeatLoader color={theme.color.secondary.default} />
        </FlexColumn>
      );
    }
    return (
      <StripeProvider stripe={stripe.stripeObject}>
        <Elements>
          <FormBilling
            stripeRef={stripeRef}
            heightRef={heightRef}
            rowWidthIndent={rowWidthIndent}
            headerWidthBufferRight={headerWidthBufferRight}
            mediaWidth={mediaWidth}
            rowHeight={rowHeight}
            rowHeightBuffer={rowHeightBuffer}
            currentFormBool={currentForm === formNames.BILLING}
            editForm={editForm}
            inputs={inputs}
            handleInputChange={this.handleInputChange}
          />
        </Elements>
      </StripeProvider>
    );
  }
}

FormContainerBilling.propTypes = {
  stripeRef: PropTypes.func.isRequired,
  heightRef: PropTypes.func.isRequired,
  stripe: PropTypes.object,
  rowWidthIndent: PropTypes.number.isRequired,
  headerWidthBufferRight: PropTypes.number.isRequired,
  mediaWidth: PropTypes.number.isRequired,
  height: PropTypes.number.isRequired,
  rowHeight: PropTypes.number.isRequired,
  rowHeightBuffer: PropTypes.number.isRequired,
  dataAccountFields: PropTypes.object.isRequired,
  statusAccount: PropTypes.object.isRequired,
  currentForm: PropTypes.string.isRequired,
  editForm: PropTypes.bool.isRequired,
  inputs: PropTypes.object.isRequired,
  setHeight: PropTypes.func.isRequired,
  clearElement: PropTypes.func.isRequired,
  resetStripeElement: PropTypes.func.isRequired,
  handleInputValueError: PropTypes.func.isRequired,
};

FormContainerBilling.defaultProps = {
  stripe: {},
};

const mapStateToProps = (state, ownProps) => ({
  stripeRef: ownProps.stripeRef,
  heightRef: ownProps.heightRef,
  stripe: state.ui.app.stripe,
  rowWidthIndent: state.ui.internal.account.utilities.forms.rowWidthIndent,
  headerWidthBufferRight: state.ui.internal.account.utilities.forms.headerWidthBufferRight,
  mediaWidth: state.ui.internal.account.utilities.pageBody.mediaWidth,
  height: state.ui.internal.account.forms[formNames.BILLING].height,
  rowHeight: state.ui.internal.account.utilities.forms.rowHeight,
  rowHeightBuffer: state.ui.internal.account.utilities.forms.rowHeightBuffer,
  heightBuffer: state.ui.internal.account.utilities.forms.heightBuffer,
  dataAccountFields: state.data.account.fields,
  statusAccount: state.data.account.status,
  currentForm: state.ui.internal.account.forms.current,
  editForm: state.ui.internal.account.forms[formNames.BILLING].edit,
  inputs: state.ui.internal.account.forms[formNames.BILLING].inputs,
  setHeight: ownProps.setHeight,
  clearElement: ownProps.clearElement,
  resetStripeElement: ownProps.resetStripeElement,
});

const mapDispatchToProps = dispatch => ({
  handleInputValueError: payload => dispatch(setInputValueError(payload)),
});

export default connect(mapStateToProps, mapDispatchToProps)(FormContainerBilling);
