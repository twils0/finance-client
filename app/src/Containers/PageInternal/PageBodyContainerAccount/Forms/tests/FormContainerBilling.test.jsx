import React from 'react';
import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';
import { StripeProvider, Elements, injectStripe } from 'react-stripe-elements';

import { initialState as initialStateAccount } from '../../../../../Reducers/dataReducersAccount';
import { initialState as initialStateUIApp } from '../../../../../Reducers/uiReducersApp';
import { initialState as initialStateUIAccount } from '../../../../../Reducers/uiReducersAccount';
import { requestStatusTypes } from '../../../../../Constants/universalConstants';
import { statusNames, fieldNames } from '../../../../../Constants/dataConstantsAccount';
import { errorMessages } from '../../../../../Constants/uiConstantsApp';
import {
  actionTypes as actionTypesUIAccount,
  formNames,
  inputNames,
} from '../../../../../Constants/uiConstantsAccount';

import FormContainerBilling from '../FormContainerBilling';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

jest.mock('react-stripe-elements', () => ({
  StripeProvider: jest.fn(() => <div />),
  Elements: jest.fn(() => <div />),
  injectStripe: jest.fn(element => element),
}));

const stripe = { stripeObject: { test: 'testStripe' } };
const stripeRef = jest.fn();
const heightRef = jest.fn();
const setHeight = jest.fn();
const clearElement = jest.fn();
const resetStripeElement = jest.fn();

const nameOnCard = 'testNameOnCard';
const promoCode = 'testPromoCode';

const shallowComponent = store =>
  shallow(
    <FormContainerBilling
      stripeRef={stripeRef}
      heightRef={heightRef}
      setHeight={setHeight}
      clearElement={clearElement}
      resetStripeElement={resetStripeElement}
    />,
    {
      context: {
        store,
      },
    },
  ).dive();

describe('Containers', () => {
  describe('PageInternal', () => {
    describe('PageBodyContainerAccount', () => {
      describe('Forms', () => {
        describe('FormContainerBilling', () => {
          afterEach(() => {
            StripeProvider.mockReset();
            Elements.mockReset();
            stripeRef.mockReset();
            heightRef.mockReset();
            setHeight.mockReset();
            clearElement.mockReset();
            resetStripeElement.mockReset();
            injectStripe.mockReset();
          });

          it('shallow renders correctly', async () => {
            const stateBeforeAccount = JSON.parse(JSON.stringify(initialStateAccount));
            const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
            const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
            const { forms } = stateBeforeUIAccount;

            stateBeforeAccount.fields[fieldNames.NAME_ON_CARD].value = nameOnCard;
            stateBeforeAccount.fields[fieldNames.PROMO_CODE].value = promoCode;

            stateBeforeUIApp.stripe = stripe;

            const store = mockStore({
              data: {
                account: stateBeforeAccount,
              },
              ui: {
                app: stateBeforeUIApp,
                internal: {
                  account: stateBeforeUIAccount,
                },
              },
            });

            const uiNameOnCard = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));
            const uiPromoCode = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));

            uiNameOnCard.value = nameOnCard;
            uiPromoCode.value = promoCode;

            const expectedActions = [
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCard,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCode,
              },
            ];

            const wrapper = shallowComponent(store);
            const stripeProviderProps = wrapper.find(StripeProvider).props();

            const actions = store.getActions();

            expect(actions).toEqual(expectedActions);
            expect(wrapper).toMatchSnapshot();
            expect(setHeight).toBeCalledWith(formNames.BILLING);
            expect(stripeProviderProps.stripe).toEqual(stripe.stripeObject);
          });

          it('shallow renders correctly, with expired promo code, componentWillUpdate, componentDidUpdate, and componentWillUnmount', async () => {
            const stateBeforeAccount = JSON.parse(JSON.stringify(initialStateAccount));
            const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
            const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
            const { forms } = stateBeforeUIAccount;

            const wrongPromoCode = 'wrongPromoCode';

            stateBeforeAccount.fields[fieldNames.NAME_ON_CARD].value = nameOnCard;
            stateBeforeAccount.fields[fieldNames.PROMO_CODE].value = wrongPromoCode;
            stateBeforeAccount.fields[fieldNames.PROMO_CODE_VALID].value = false;

            stateBeforeUIApp.stripe = stripe;

            const store = mockStore({
              data: {
                account: stateBeforeAccount,
              },
              ui: {
                app: stateBeforeUIApp,
                internal: {
                  account: stateBeforeUIAccount,
                },
              },
            });

            const uiNameOnCard = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));
            const uiPromoCode = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));

            uiNameOnCard.value = nameOnCard;
            uiPromoCode.value = wrongPromoCode;
            uiPromoCode.errorMessage = errorMessages.PROMO_CODE_EXPIRED;

            const nextProps = {
              statusAccount: {
                [statusNames.UPDATE_STRIPE_FIELDS_REQUEST]: {
                  status: requestStatusTypes.LOADING,
                },
              },
              clearElement,
              resetStripeElement,
            };

            const expectedActions = [
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCard,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCode,
              },
            ];

            const wrapper = shallowComponent(store);
            const stripeProviderProps = wrapper.find(StripeProvider).props();

            const instance = wrapper.instance();

            instance.componentWillUpdate(nextProps);
            instance.componentDidUpdate();
            instance.componentWillUnmount();

            const actions = store.getActions();

            expect(actions).toEqual(expectedActions);
            expect(wrapper).toMatchSnapshot();
            expect(setHeight).toBeCalledWith(formNames.BILLING);
            expect(clearElement).toBeCalledWith(formNames.BILLING);
            expect(resetStripeElement).toBeCalled();
            expect(setHeight).toBeCalledWith(formNames.BILLING);
            expect(clearElement).toBeCalledWith(formNames.BILLING);
            expect(stripeProviderProps.stripe).toEqual(stripe.stripeObject);
            expect(setHeight).toHaveBeenCalledTimes(2);
            expect(clearElement).toHaveBeenCalledTimes(2);
            expect(resetStripeElement).toHaveBeenCalledTimes(1);
          });

          it('shallow renders correctly, with loading indicator', async () => {
            const stateBeforeAccount = JSON.parse(JSON.stringify(initialStateAccount));
            const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
            const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
            const { forms } = stateBeforeUIAccount;

            stateBeforeAccount.status[statusNames.UPDATE_STRIPE_FIELDS_REQUEST].status =
              requestStatusTypes.LOADING;

            stateBeforeAccount.fields[fieldNames.NAME_ON_CARD].value = nameOnCard;
            stateBeforeAccount.fields[fieldNames.PROMO_CODE].value = promoCode;

            stateBeforeUIApp.stripe = stripe;

            const store = mockStore({
              data: {
                account: stateBeforeAccount,
              },
              ui: {
                app: stateBeforeUIApp,
                internal: {
                  account: stateBeforeUIAccount,
                },
              },
            });

            const uiNameOnCard = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));
            const uiPromoCode = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));

            uiNameOnCard.value = nameOnCard;
            uiPromoCode.value = promoCode;

            const expectedActions = [
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCard,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCode,
              },
            ];

            const wrapper = shallowComponent(store);

            const actions = store.getActions();

            expect(actions).toEqual(expectedActions);
            expect(wrapper).toMatchSnapshot();
          });

          it('shallow renders correctly and calls handleInputChange, with no name on card error', async () => {
            const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
            const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
            const { forms } = stateBeforeUIAccount;

            stateBeforeUIApp.stripe = stripe;

            forms[formNames.BILLING].inputs[
              inputNames[formNames.BILLING].NAME_ON_CARD
            ].errorMessage =
              errorMessages.NO_NAME_ON_CARD;

            const store = mockStore({
              data: {
                account: initialStateAccount,
              },
              ui: {
                app: stateBeforeUIApp,
                internal: {
                  account: stateBeforeUIAccount,
                },
              },
            });

            const uiNameOnCard = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));
            const uiPromoCode = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));
            const uiNameOnCardChange = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));

            uiNameOnCardChange.value = nameOnCard;
            uiNameOnCardChange.errorMessage = '';

            const inputEvent = {
              preventDefault: jest.fn(),
              target: {
                id: inputNames[formNames.BILLING].NAME_ON_CARD,
                value: nameOnCard,
              },
            };
            const expectedActions = [
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCard,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCode,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCardChange,
              },
            ];

            const wrapper = shallowComponent(store);

            const instance = wrapper.instance();

            instance.handleInputChange(inputEvent);

            const actions = store.getActions();

            expect(actions).toEqual(expectedActions);
            expect(wrapper).toMatchSnapshot();
          });

          it('shallow renders correctly and calls handleInputChange, with promo code invalid error', async () => {
            const stateBeforeAccount = JSON.parse(JSON.stringify(initialStateAccount));
            const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
            const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
            const { forms } = stateBeforeUIAccount;

            const wrongPromoCode = 'wrongPromoCode';

            stateBeforeAccount.fields[fieldNames.PROMO_CODE].value = wrongPromoCode;

            stateBeforeUIApp.stripe = stripe;

            forms[formNames.BILLING].inputs[inputNames[formNames.BILLING].PROMO_CODE].errorMessage =
              errorMessages.PROMO_CODE_INVALID;

            const store = mockStore({
              data: {
                account: stateBeforeAccount,
              },
              ui: {
                app: stateBeforeUIApp,
                internal: {
                  account: stateBeforeUIAccount,
                },
              },
            });

            const uiNameOnCard = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));
            const uiPromoCode = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));
            const uiPromoCodeChange = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));

            uiPromoCode.value = wrongPromoCode;

            uiPromoCodeChange.value = promoCode;
            uiPromoCodeChange.errorMessage = '';

            const inputEvent = {
              preventDefault: jest.fn(),
              target: {
                id: inputNames[formNames.BILLING].PROMO_CODE,
                value: promoCode,
              },
            };
            const expectedActions = [
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCard,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCode,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCodeChange,
              },
            ];

            const wrapper = shallowComponent(store);

            const instance = wrapper.instance();

            instance.handleInputChange(inputEvent);

            const actions = store.getActions();

            expect(actions).toEqual(expectedActions);
            expect(wrapper).toMatchSnapshot();
          });

          it('shallow renders correctly and calls handleInputChange, with promo code expired error', async () => {
            const stateBeforeAccount = JSON.parse(JSON.stringify(initialStateAccount));
            const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
            const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
            const { forms } = stateBeforeUIAccount;

            const wrongPromoCode = 'wrongPromoCode';

            stateBeforeAccount.fields[fieldNames.PROMO_CODE].value = wrongPromoCode;

            stateBeforeUIApp.stripe = stripe;

            forms[formNames.BILLING].inputs[inputNames[formNames.BILLING].PROMO_CODE].errorMessage =
              errorMessages.PROMO_CODE_EXPIRED;

            const store = mockStore({
              data: {
                account: stateBeforeAccount,
              },
              ui: {
                app: stateBeforeUIApp,
                internal: {
                  account: stateBeforeUIAccount,
                },
              },
            });

            const uiNameOnCard = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));
            const uiPromoCode = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));
            const uiPromoCodeChange = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));

            uiPromoCode.value = wrongPromoCode;

            uiPromoCodeChange.value = promoCode;
            uiPromoCodeChange.errorMessage = '';

            const inputEvent = {
              preventDefault: jest.fn(),
              target: {
                id: inputNames[formNames.BILLING].PROMO_CODE,
                value: promoCode,
              },
            };
            const expectedActions = [
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCard,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCode,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCodeChange,
              },
            ];

            const wrapper = shallowComponent(store);

            const instance = wrapper.instance();

            instance.handleInputChange(inputEvent);

            const actions = store.getActions();

            expect(actions).toEqual(expectedActions);
            expect(wrapper).toMatchSnapshot();
          });

          it('shallow renders correctly and calls handleInputChange, with undefined error', async () => {
            const stateBeforeAccount = JSON.parse(JSON.stringify(initialStateAccount));
            const stateBeforeUIApp = JSON.parse(JSON.stringify(initialStateUIApp));
            const stateBeforeUIAccount = JSON.parse(JSON.stringify(initialStateUIAccount));
            const { forms } = stateBeforeUIAccount;

            stateBeforeUIApp.stripe = stripe;

            forms[formNames.BILLING].inputs[inputNames[formNames.BILLING].STRIPE].value =
              errorMessages.INTERNAL_ERROR;
            forms[formNames.BILLING].inputs[inputNames[formNames.BILLING].STRIPE].errorMessage =
              errorMessages.INTERNAL_ERROR;

            const store = mockStore({
              data: {
                account: stateBeforeAccount,
              },
              ui: {
                app: stateBeforeUIApp,
                internal: {
                  account: stateBeforeUIAccount,
                },
              },
            });

            const uiNameOnCard = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].NAME_ON_CARD]));
            const uiPromoCode = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].PROMO_CODE]));
            const uiStripe = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].STRIPE]));
            const uiStripeChange = JSON.parse(JSON.stringify(forms[formNames.BILLING]
              .inputs[inputNames[formNames.BILLING].STRIPE]));

            uiStripe.value = errorMessages.INTERNAL_ERROR;

            uiStripeChange.value = errorMessages.INTERNAL_ERROR;
            uiStripeChange.errorMessage = '';

            const inputEvent = {
              preventDefault: jest.fn(),
              target: {
                id: inputNames[formNames.BILLING].STRIPE,
                value: errorMessages.INTERNAL_ERROR,
              },
            };
            const expectedActions = [
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiNameOnCard,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiPromoCode,
              },
              {
                type: actionTypesUIAccount.SET_INPUT_VALUE_ERROR,
                payload: uiStripeChange,
              },
            ];

            const wrapper = shallowComponent(store);

            const instance = wrapper.instance();

            instance.handleInputChange(inputEvent);

            const actions = store.getActions();

            expect(actions).toEqual(expectedActions);
            expect(wrapper).toMatchSnapshot();
          });
        });
      });
    });
  });
});
