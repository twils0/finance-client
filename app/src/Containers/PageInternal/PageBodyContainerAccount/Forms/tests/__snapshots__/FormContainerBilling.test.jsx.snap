// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers PageInternal PageBodyContainerAccount Forms FormContainerBilling shallow renders correctly 1`] = `
<mockConstructor
  stripe={
    Object {
      "test": "testStripe",
    }
  }
>
  <mockConstructor>
    <FormBilling
      currentFormBool={false}
      editForm={false}
      handleInputChange={[Function]}
      headerWidthBufferRight={10}
      heightRef={[MockFunction]}
      inputs={
        Object {
          "billingNameOnCard": Object {
            "errorMessage": "",
            "id": "billingNameOnCard",
            "parentId": "billingForm",
            "value": "testNameOnCard",
          },
          "billingPromoCode": Object {
            "errorMessage": "",
            "id": "billingPromoCode",
            "parentId": "billingForm",
            "value": "testPromoCode",
          },
          "billingStripe": Object {
            "errorMessage": "",
            "id": "billingStripe",
            "parentId": "billingForm",
            "value": "",
          },
        }
      }
      mediaWidth={874}
      rowHeight={60}
      rowHeightBuffer={18}
      rowWidthIndent={10}
      stripeRef={[MockFunction]}
    />
  </mockConstructor>
</mockConstructor>
`;

exports[`Containers PageInternal PageBodyContainerAccount Forms FormContainerBilling shallow renders correctly and calls handleInputChange, with no name on card error 1`] = `
<mockConstructor
  stripe={
    Object {
      "test": "testStripe",
    }
  }
>
  <mockConstructor>
    <FormBilling
      currentFormBool={false}
      editForm={false}
      handleInputChange={[Function]}
      headerWidthBufferRight={10}
      heightRef={[MockFunction]}
      inputs={
        Object {
          "billingNameOnCard": Object {
            "errorMessage": "Please enter the name shown on your credit card.",
            "id": "billingNameOnCard",
            "parentId": "billingForm",
            "value": "",
          },
          "billingPromoCode": Object {
            "errorMessage": "",
            "id": "billingPromoCode",
            "parentId": "billingForm",
            "value": "",
          },
          "billingStripe": Object {
            "errorMessage": "",
            "id": "billingStripe",
            "parentId": "billingForm",
            "value": "",
          },
        }
      }
      mediaWidth={874}
      rowHeight={60}
      rowHeightBuffer={18}
      rowWidthIndent={10}
      stripeRef={[MockFunction]}
    />
  </mockConstructor>
</mockConstructor>
`;

exports[`Containers PageInternal PageBodyContainerAccount Forms FormContainerBilling shallow renders correctly and calls handleInputChange, with promo code expired error 1`] = `
<mockConstructor
  stripe={
    Object {
      "test": "testStripe",
    }
  }
>
  <mockConstructor>
    <FormBilling
      currentFormBool={false}
      editForm={false}
      handleInputChange={[Function]}
      headerWidthBufferRight={10}
      heightRef={[MockFunction]}
      inputs={
        Object {
          "billingNameOnCard": Object {
            "errorMessage": "",
            "id": "billingNameOnCard",
            "parentId": "billingForm",
            "value": "",
          },
          "billingPromoCode": Object {
            "errorMessage": "This promo code is no longer valid.",
            "id": "billingPromoCode",
            "parentId": "billingForm",
            "value": "wrongPromoCode",
          },
          "billingStripe": Object {
            "errorMessage": "",
            "id": "billingStripe",
            "parentId": "billingForm",
            "value": "",
          },
        }
      }
      mediaWidth={874}
      rowHeight={60}
      rowHeightBuffer={18}
      rowWidthIndent={10}
      stripeRef={[MockFunction]}
    />
  </mockConstructor>
</mockConstructor>
`;

exports[`Containers PageInternal PageBodyContainerAccount Forms FormContainerBilling shallow renders correctly and calls handleInputChange, with promo code invalid error 1`] = `
<mockConstructor
  stripe={
    Object {
      "test": "testStripe",
    }
  }
>
  <mockConstructor>
    <FormBilling
      currentFormBool={false}
      editForm={false}
      handleInputChange={[Function]}
      headerWidthBufferRight={10}
      heightRef={[MockFunction]}
      inputs={
        Object {
          "billingNameOnCard": Object {
            "errorMessage": "",
            "id": "billingNameOnCard",
            "parentId": "billingForm",
            "value": "",
          },
          "billingPromoCode": Object {
            "errorMessage": "Please enter a valid promo code.",
            "id": "billingPromoCode",
            "parentId": "billingForm",
            "value": "wrongPromoCode",
          },
          "billingStripe": Object {
            "errorMessage": "",
            "id": "billingStripe",
            "parentId": "billingForm",
            "value": "",
          },
        }
      }
      mediaWidth={874}
      rowHeight={60}
      rowHeightBuffer={18}
      rowWidthIndent={10}
      stripeRef={[MockFunction]}
    />
  </mockConstructor>
</mockConstructor>
`;

exports[`Containers PageInternal PageBodyContainerAccount Forms FormContainerBilling shallow renders correctly and calls handleInputChange, with undefined error 1`] = `
<mockConstructor
  stripe={
    Object {
      "test": "testStripe",
    }
  }
>
  <mockConstructor>
    <FormBilling
      currentFormBool={false}
      editForm={false}
      handleInputChange={[Function]}
      headerWidthBufferRight={10}
      heightRef={[MockFunction]}
      inputs={
        Object {
          "billingNameOnCard": Object {
            "errorMessage": "",
            "id": "billingNameOnCard",
            "parentId": "billingForm",
            "value": "",
          },
          "billingPromoCode": Object {
            "errorMessage": "",
            "id": "billingPromoCode",
            "parentId": "billingForm",
            "value": "",
          },
          "billingStripe": Object {
            "errorMessage": "Please try your request again. If unsuccessful on subsequent attempts, please contact customer service.",
            "id": "billingStripe",
            "parentId": "billingForm",
            "value": "Please try your request again. If unsuccessful on subsequent attempts, please contact customer service.",
          },
        }
      }
      mediaWidth={874}
      rowHeight={60}
      rowHeightBuffer={18}
      rowWidthIndent={10}
      stripeRef={[MockFunction]}
    />
  </mockConstructor>
</mockConstructor>
`;

exports[`Containers PageInternal PageBodyContainerAccount Forms FormContainerBilling shallow renders correctly, with expired promo code, componentWillUpdate, componentDidUpdate, and componentWillUnmount 1`] = `
<mockConstructor
  stripe={
    Object {
      "test": "testStripe",
    }
  }
>
  <mockConstructor>
    <FormBilling
      currentFormBool={false}
      editForm={false}
      handleInputChange={[Function]}
      headerWidthBufferRight={10}
      heightRef={[MockFunction]}
      inputs={
        Object {
          "billingNameOnCard": Object {
            "errorMessage": "",
            "id": "billingNameOnCard",
            "parentId": "billingForm",
            "value": "testNameOnCard",
          },
          "billingPromoCode": Object {
            "errorMessage": "This promo code is no longer valid.",
            "id": "billingPromoCode",
            "parentId": "billingForm",
            "value": "wrongPromoCode",
          },
          "billingStripe": Object {
            "errorMessage": "",
            "id": "billingStripe",
            "parentId": "billingForm",
            "value": "",
          },
        }
      }
      mediaWidth={874}
      rowHeight={60}
      rowHeightBuffer={18}
      rowWidthIndent={10}
      stripeRef={[MockFunction]}
    />
  </mockConstructor>
</mockConstructor>
`;

exports[`Containers PageInternal PageBodyContainerAccount Forms FormContainerBilling shallow renders correctly, with loading indicator 1`] = `
<FlexColumn
  alignItems={null}
  alignSelf={null}
  borderColorType="border"
  borderRadius={null}
  borderStyle={null}
  borderWidth={null}
  boxSizing={null}
  color={Object {}}
  display={null}
  flex={null}
  flexWrap={null}
  height="0px"
  justifyContent={null}
  margin={null}
  mediaWidth={null}
  minHeight={null}
  minWidth={null}
  overflowX={null}
  overflowY={null}
  padding={null}
  position={null}
  responsiveAlignSecurities={null}
  responsiveJustifyContent={null}
  responsiveMargin={null}
  responsivePadding={null}
  responsivePosition={null}
  responsiveWidth={null}
  separator={false}
  themeColor="primary"
  width="100%"
>
  <BeatLoader
    color="#7367e6"
    loading={true}
    margin="2px"
    size="15px"
  />
</FlexColumn>
`;
